Automattic/vip-go-mu-plugins
# Use wpcf7_upload_dir filter to set Contact Form 7 uploads directory

## Description

In 2021, Contact Form 7 changed the way it handles the `WPCF7_UPLOADS_TMP_DIR` constant (see https://github.com/rocklobster-in/contact-form-7/commit/90e87ab044ce6b4ff4e4df755576795908340181), which resulted in [our fix](https://github.com/Automattic/vip-go-mu-plugins/blob/524a46de14a6e5b0bc733348c1866b8a344468a5/plugin-fixes.php#L10-L19) no longer working.

The issue is that the plugin only returns `WPCF7_UPLOADS_TMP_DIR` [when the value is inside `WP_CONTENT_DIR`](https://github.com/rocklobster-in/contact-form-7/blob/2cfaa472fa485c6d3366fcdd80701fdaf7f9e425/includes/file.php#L322). When it's not (as in our case where we set it to `/tmp/cf7`), the plugin ignores the value of `WPCF7_UPLOADS_TMP_DIR` and [instead uses its own `wpcf7_upload_dir()` function](https://github.com/rocklobster-in/contact-form-7/blob/2cfaa472fa485c6d3366fcdd80701fdaf7f9e425/includes/file.php#L327) to build the uploads path, setting it to a path inside `wp-content/` and trying to write there, which fails on the VIP platform due to that path being an object store.

So, setting `WPCF7_UPLOADS_TMP_DIR` to `/tmp/cf7` is ineffective. Instead, we need to filter `wpcf7_upload_dir` to ensure that Contact Form 7 only attempts to write to `/tmp/cf7`.

This PR filters `wpcf7_upload_dir` and sets the directory to `/tmp/cf7/`, ensuring that Contact Form 7 will only use that temp path for uploads and not try to write to `wp-content/`.

This fix should allow us to remove [all of the existing Contact Form 7 fixes](https://github.com/Automattic/vip-go-mu-plugins/blob/524a46de14a6e5b0bc733348c1866b8a344468a5/plugin-fixes.php#L10-L42) we currently have in the VIP MU plugins:

- [`WPCF7_UPLOADS_TMP_DIR` does not need to be set](https://github.com/Automattic/vip-go-mu-plugins/blob/524a46de14a6e5b0bc733348c1866b8a344468a5/plugin-fixes.php#L10C4-L19) because the plugin ignores it anyway when we set it to a `/tmp/ path and if we're filtering `wpcf7_upload_dir`, there's no need to also define `WPCF7_UPLOADS_TMP_DIR`.
- [The fix for cleaning up uploaded attachments](https://github.com/Automattic/vip-go-mu-plugins/blob/524a46de14a6e5b0bc733348c1866b8a344468a5/plugin-fixes.php#L21-L42) can also be removed, since [`wpcf7_upload_tmp_dir()` will always return our temp directory](https://github.com/rocklobster-in/contact-form-7/blob/2cfaa472fa485c6d3366fcdd80701fdaf7f9e425/includes/file.php#L347) if we're filtering `wpcf7_upload_dir`, so there should be no need to remove the `wpcf7_cleanup_upload_files` action (this fix was described here: https://github.com/Automattic/vip-go-mu-plugins/pull/4044)

I have not removed those fixes in this PR. They are redundant with this filter in place, but they do not cause any conflict with added filter.



## Changelog Description

### Plugin fixes updated to improve Contact Form 7 upload compatibility 

Added a filter to set the Contact Form 7 uploads directory to a temp directory.



## Pre-review checklist

Please make sure the items below have been covered before requesting a review:


## Pre-deploy checklist


## Steps to Test
1. Install a recent Contact Form 7 (e.g. 5.8) on a VIP test site
2. Create a contact form that includes a field for uploading an attachment
3. Submit the form with the attachment and notice how the form fails to submit due to CF7 trying to write to `wp-content/`
4. Checkout this PR and switch the site to use the appropriate version of mu-plugins
5. Perform the same test with submitting a form that includes an attachment and observe that the form submits as expected as CF7 writes to the temp directory instead of `wp-content/`
